### PR TITLE
introduce tax/discount labels and handle tds flow (additon, removal) on update and create of invoice

### PR DESCRIPTION
**Pull request by ahmad19**

### CODE REVIEW

File: app/controllers/invoices_controller.rb [hunk 1]
```diff
@@ -19,43 +19,43 @@ def new
 
   # GET /invoices/1/edit
   def edit
-   @invoice.line_items.each do |line_item|
-    if line_item.tax_and_discount_polies.empty?
-      line_item.tax_and_discount_polies.build
-    end
-   end
+  #   @invoice.line_items.each do |line_item|
```
Comment (Azher9767) [inline app/controllers/invoices_controller.rb:22]: Remove this commented code.

File: app/controllers/invoices_controller.rb [hunk 2]
```diff
@@ -82,22 +82,27 @@ def set_category
 
   # Only allow a list of trusted parameters through.
   def invoice_params
-    params[:invoice][:line_items_attributes].each do |_key, value|
-      value[:tax_and_discount_polies_attributes] = value[:tax_and_discount_polies_attributes].each do |_k, v|
-        v[:tax_and_discount_id] = v[:tax_and_discount_id].reject!(&:empty?)
-      end
-    end if params[:invoice][:line_items_attributes].present?
+    # params[:invoice][:line_items_attributes].each do |_key, value|
```
Comment (Azher9767) [inline app/controllers/invoices_controller.rb:85]: Remove this commented code

File: app/helpers/invoices_helper.rb [hunk 3]
```diff
@@ -8,12 +8,14 @@ def categories(invoice)
...
+  def td_options(current_user, object)
+    selected = object.tax_and_discount_polies.map(&:tax_and_discount_id)
+    selected_ids = object.tax_and_discount_polies.map { |tdp| { id: tdp.tax_and_discount_id, name: tdp.name, p_id: tdp.id }}
+
     tds = current_user.tax_and_discounts
     [
-      ["Taxes", tds.tax.map{|t| [t.name, t.id, { selected: selected.include?(t.name) }]}],
-      ["Discount", tds.discount.map{|d| [d.name, d.id, { selected: selected.include?(d.name) }]}]
+      ['Taxes', tds.tax.map { |t| [t.name, t.id, { selected: selected.include?(t.id), data: { poly_id: selected_ids.select {|td| td[:id] == t.id }} }] }],
+      ['Discount', tds.discount.map { |d| [d.name, d.id, { selected: selected.include?(d.id), data: { poly_id: selected_ids.select {|td| td[:id] == d.id }} }] }]
```
Comment (Azher9767) [inline app/helpers/invoices_helper.rb:18]: Create a ticket to refactor this.

File: app/javascript/controllers/invoice_controller.js [hunk 4]
```diff
@@ -112,90 +112,100 @@ export default class extends Controller {
...
+    // this.removeLineItemTdsFromDb(event)
     this.taxAndDiscountOfLineItems();
   }
 
-  taxAndDiscountOfLineItems(){
-    let lineItemsAttributes = [];
-    this.lineitemTargets.forEach((lineitem) => {
-      const selectedTds = Array.from(lineitem.children[4].children[2].children[0].children).slice(0, -1)
-      const tdIds = []
+  removeLineItemTdsFromDb(event) {
```
Comment (Azher9767) [inline app/javascript/controllers/invoice_controller.js:154]: unused method.

File: app/javascript/controllers/invoice_controller.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import { Controller } from "@hotwired/stimulus"
+import { Controller, ElementObserver } from "@hotwired/stimulus"
```
Comment (Azher9767) [inline app/javascript/controllers/invoice_controller.js:1]: Figure out ununsed imports and remove them.

File: app/services/calculation/line_items_with_td_and_itd.rb [hunk 6]
```diff
@@ -33,29 +33,43 @@ def sub_total
 
     def apply_polies(amount, policies)
       policies.reduce(amount) do |current_amount, td|
-        current_amount += current_amount * td.amount / 100.0
-        current_amount
+        if td.discount? && !td.marked_for_destruction?
```
Comment (Azher9767) [inline app/services/calculation/line_items_with_td_and_itd.rb:36]: Lets see if it is possible to reject the "to be" deleted objects are rejected from the source.

File: app/services/invoice_object_builder.rb [hunk 7]
```diff
@@ -1,7 +1,8 @@
+# frozen_string_literal: true
```
Comment (Azher9767) [inline app/services/invoice_object_builder.rb:1]: we don't need this class.

File: app/controllers/invoices/calculations_controller.rb [hunk 8]
```diff
@@ -40,7 +44,7 @@ def create # rubocop:disable Metrics/AbcSize,Metrics/MethodLength
     def calculation_params
       params.require(:calculation).permit(
         :taxAndDiscountPolyAttributes,
-        lineItemsAttributes: [:quantity, :unitRate, :unit, { tdIds: [] }],
+        lineItemsAttributes: [:quantity, :unitRate, :unit, :objId, { tdIds: [] }],
```
Comment (ahmad19) [inline app/controllers/invoices/calculations_controller.rb:47]: we should mention the purpose of this `objId` attribute just above this `calculation_params`

File: app/javascript/controllers/invoice_controller.js [hunk 9]
```diff
@@ -112,90 +112,80 @@ export default class extends Controller {
...
-      rowHtmlId = `#invoice_line_items_attributes_${rowIdOther}_tax_and_discount_polies_attributes_0_tax_and_discount_id`;
+      rowHtmlId = `#invoice_line_items_attributes_${rowIdOther}_tax_and_discount_ids`;
     }
   
     new TomSelect(rowHtmlId, { 
       plugins: ['checkbox_options', 'remove_button'],
       create: false,
       hidePlaceholder: true,
       onItemAdd: addEventHandlerTd(),
+      onItemRemove:removeEventHandlerTd(),
```
Comment (ahmad19) [inline app/javascript/controllers/invoice_controller.js:130]: space

File: app/services/invoice_object_builder_for_create.rb [hunk 10]
```diff
@@ -0,0 +1,47 @@
+# This class is responsible for building the invoice object
+# It will take the invoice params and return the invoice object
+# It will also handle the validation of the invoice params
```
Comment (ahmad19) [inline app/services/invoice_object_builder_for_create.rb:3]: This class's whole purpose is to manually populate polies based on the tax/discount selected.

File: app/services/line_items_tds_processor.rb [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+  def build_polies(line_item, td_ids)
+    TaxAndDiscount.where(id: td_ids).each do |td|
+      line_item.tax_and_discount_polies.build(name: td.name, td_type: td.td_type, amount: td.amount, tax_and_discount_id: td.id,
+                                              tax_discountable: line_item)
+    end
+  end
+
+  def populate_polies(line_item, td_ids)
+    tds = TaxAndDiscount.where(id: td_ids)
```
Comment (ahmad19) [inline app/services/line_items_tds_processor.rb:87]: Is it possible to move out the TaxAndDiscount queries out of this place at one global location ?

File: spec/features/category_spec.rb [hunk 12]
```diff
@@ -3,22 +3,22 @@
...
 
-    expect(page).to have_content('Category was successfully created.')
+    expect(page.has_content?('Category was successfully created.')).to be(true)
     category = Category.last
-    expect(page).to have_content("Name: #{category.name}")
-    expect(page).to have_content("User: #{category.user_id}")
-    expect(category.product_count).to eq(1) 
+    expect(page.has_content?("Name: #{category.name}")).to be(true)
+    expect(page.has_content?("User: #{category.user_id}")).to be(true)
+    expect(category.product_count).to eq(1)
```
Comment (ahmad19) [inline spec/features/category_spec.rb:22]: `to be_truthy`

File: spec/features/invoice_spec.rb [hunk 13]
```diff
@@ -8,39 +8,39 @@
...
-    expect(page).to have_content("Payment date: #{invoice.payment_date}")
-    expect(page).to have_content("Due date: #{invoice.due_date}")
+    expect(page.has_content?("User: #{invoice.user_id}")).to be(true)
+    expect(page.has_content?("Sub total: #{invoice.sub_total}")).to be(true)
+    expect(page.has_content?("Name: #{invoice.name}")).to be(true)
+    expect(page.has_content?("Status: #{invoice.status}")).to be(true)
+    expect(page.has_content?(invoice.line_items_count.to_s)).to be(true)
+    expect(page.has_content?("Note: #{invoice.note}")).to be(true)
+    expect(page.has_content?("Payment date: #{invoice.payment_date}")).to be(true)
+    expect(page.has_content?("Due date: #{invoice.due_date}")).to be(true)
```
Comment (ahmad19) [inline spec/features/invoice_spec.rb:41]: same, use `be_truthy`

File: spec/features/product_spec.rb [hunk 14]
```diff
@@ -3,26 +3,28 @@
...
-    expect(page).to have_content("Name: #{product.name}")
-    expect(page).to have_content("Description: #{product.description}")
-    expect(page).to have_content("Unit rate: #{product.unit_rate}")
-    expect(page).to have_content("Category: #{product.category_id}")
-    expect(page).to have_content("Unit: #{product.unit}")
+    expect(page.has_content?("Name: #{product.name}")).to be(true)
+    expect(page.has_content?("Description: #{product.description}")).to be(true)
+    expect(page.has_content?("Unit rate: #{product.unit_rate}")).to be(true)
+    expect(page.has_content?("Category: #{product.category_id}")).to be(true)
+    expect(page.has_content?("Unit: #{product.unit}")).to be(true)
```
Comment (ahmad19) [inline spec/features/product_spec.rb:28]: Same, use `be_truthy`

**Review by ahmad19 — COMMENTED** [review]
LGTM

**Pull request merged**